File: Treachery-iOS/Treachery-iOS/Home/LobbyViewModel.swift
//
//  LobbyViewModel.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/13/26.
//

import Foundation
import SwiftUI

@MainActor
final class LobbyViewModel: ObservableObject {
    @Published var game: Game?
    @Published var players: [Player] = []
    @Published var errorMessage: String?
    @Published var isStartingGame = false
    @Published var isGameDisbanded = false

    let gameId: String
    let isHost: Bool
    var currentUserId: String?
    private let firestoreManager: FirestoreManaging
    private let cloudFunctions: CloudFunctionsProtocol
    private var gameListener: ListenerCancellable?
    private var playersListener: ListenerCancellable?
    private var hasReceivedFirstSnapshot = false

    var currentPlayer: Player? {
        guard let userId = currentUserId else { return nil }
        return players.first { $0.userId == userId }
    }

    var canStartGame: Bool {
        guard let game = game else { return false }
        let minPlayers = game.gameMode.includesTreachery ? Role.minimumPlayerCount : 1
        return isHost && players.count >= minPlayers
    }

    var minimumPlayerCount: Int {
        guard let game = game else { return Role.minimumPlayerCount }
        return game.gameMode.includesTreachery ? Role.minimumPlayerCount : 1
    }

    var isGameStarted: Bool {
        game?.state == .inProgress
    }

    init(
        gameId: String,
        isHost: Bool,
        firestoreManager: FirestoreManaging = FirestoreManager(),
        cloudFunctions: CloudFunctionsProtocol = CloudFunctions()
    ) {
        self.gameId = gameId
        self.isHost = isHost
        self.firestoreManager = firestoreManager
        self.cloudFunctions = cloudFunctions
        startListening()
    }

    deinit {
        gameListener?.remove()
        playersListener?.remove()
    }

    // MARK: - Listeners

    private func startListening() {
        gameListener = firestoreManager.listenToGame(id: gameId) { [weak self] game in
            Task { @MainActor in
                guard let self else { return }
                if game == nil && self.hasReceivedFirstSnapshot {
                    // Game was deleted (host left)
                    self.isGameDisbanded = true
                }
                self.game = game
                self.hasReceivedFirstSnapshot = true
            }
        }
        playersListener = firestoreManager.listenToPlayers(gameId: gameId) { [weak self] players in
            Task { @MainActor in
                withAnimation(.easeInOut(duration: 0.3)) {
                    self?.players = players
                }
            }
        }
    }

    private func stopListening() {
        gameListener?.remove()
        playersListener?.remove()
    }

    // MARK: - Actions

    func startGame() async {
        guard isHost else { return }
        errorMessage = nil
        isStartingGame = true
        do {
            try await cloudFunctions.startGame(gameId: gameId)
            AnalyticsService.trackEvent("start_game", params: [
                "player_count": players.count,
                "game_mode": game?.gameMode.rawValue ?? "unknown"
            ])
        } catch {
            errorMessage = error.localizedDescription
        }
        isStartingGame = false
    }

    func leaveGame(userId: String) async {
        errorMessage = nil
        // Stop listeners before leaving to prevent race conditions
        // where the snapshot update re-renders the view mid-navigation
        stopListening()
        do {
            try await cloudFunctions.leaveGame(gameId: gameId)
            AnalyticsService.trackEvent("leave_lobby")
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Player Customization

    func updatePlayerColor(_ hex: String?) async {
        guard let player = currentPlayer else { return }
        do {
            try await firestoreManager.updatePlayerColor(gameId: gameId, playerId: player.id, color: hex)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateCommanderName(_ name: String?) async {
        guard let player = currentPlayer else { return }
        do {
            try await firestoreManager.updateCommanderName(gameId: gameId, playerId: player.id, name: name)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/ui/lobby/LobbyViewModel.kt
package com.solomon.treachery.ui.lobby

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solomon.treachery.data.AnalyticsService
import com.solomon.treachery.data.CloudFunctionsRepository
import com.solomon.treachery.data.FirestoreRepository
import com.solomon.treachery.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LobbyViewModel @Inject constructor(
    private val firestoreRepository: FirestoreRepository,
    private val cloudFunctionsRepository: CloudFunctionsRepository,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    val gameId: String = savedStateHandle["gameId"] ?: ""
    private val navIsHost: Boolean = savedStateHandle["isHost"] ?: false

    private val _game = MutableStateFlow<Game?>(null)
    val game: StateFlow<Game?> = _game.asStateFlow()

    private val _players = MutableStateFlow<List<Player>>(emptyList())
    val players: StateFlow<List<Player>> = _players.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isStartingGame = MutableStateFlow(false)
    val isStartingGame: StateFlow<Boolean> = _isStartingGame.asStateFlow()

    private val _isGameDisbanded = MutableStateFlow(false)
    val isGameDisbanded: StateFlow<Boolean> = _isGameDisbanded.asStateFlow()

    // The Game model does not surface max_traitor_rarity yet, so track the
    // host's selection locally, seeded from what this session last sent the
    // server (falling back to the server default of allowing every traitor).
    private val _maxTraitorRarity = MutableStateFlow(
        cloudFunctionsRepository.lastKnownMaxTraitorRarity(gameId)
            ?.let { Rarity.fromValue(it) } ?: Rarity.SPECIAL
    )
    val maxTraitorRarity: StateFlow<Rarity> = _maxTraitorRarity.asStateFlow()

    var currentUserId: String? = null

    private var hasReceivedFirstSnapshot = false

    val currentPlayer: Player?
        get() = currentUserId?.let { uid -> players.value.find { it.userId == uid } }

    /** True when the signed-in user is the live host (updates after host promotion). */
    val isHost: Boolean
        get() {
            val uid = currentUserId ?: return navIsHost
            val hostId = game.value?.hostId
            return if (hostId != null) hostId == uid else navIsHost
        }

    val canStartGame: Boolean
        get() {
            val g = game.value ?: return false
            val minPlayers = if (g.gameMode.includesTreachery) Role.MINIMUM_PLAYER_COUNT else 1
            return isHost && players.value.size >= minPlayers
        }

    val minimumPlayerCount: Int
        get() {
            val g = game.value ?: return Role.MINIMUM_PLAYER_COUNT
            return if (g.gameMode.includesTreachery) Role.MINIMUM_PLAYER_COUNT else 1
        }

    val isGameStarted: Boolean
        get() = game.value?.state == GameState.IN_PROGRESS

    init {
        startListening()
    }

    private fun startListening() {
        viewModelScope.launch {
            firestoreRepository.gameFlow(gameId).collect { game ->
                if (game == null && hasReceivedFirstSnapshot) {
                    _isGameDisbanded.value = true
                }
                _game.value = game
                hasReceivedFirstSnapshot = true
            }
        }
        viewModelScope.launch {
            firestoreRepository.playersFlow(gameId).collect { playerList ->
                _players.value = playerList
            }
        }
    }

    fun startGame() {
        if (!isHost) return
        _errorMessage.value = null
        _isStartingGame.value = true
        viewModelScope.launch {
            try {
                cloudFunctionsRepository.startGame(gameId)
                AnalyticsService.trackEvent("start_game", mapOf(
                    "player_count" to players.value.size.toString(),
                    "game_mode" to (game.value?.gameMode?.value ?: "unknown")
                ))
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
            _isStartingGame.value = false
        }
    }

    fun leaveGame() {
        _errorMessage.value = null
        viewModelScope.launch {
            try {
                cloudFunctionsRepository.leaveGame(gameId)
                AnalyticsService.trackEvent("leave_lobby")
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

    fun updatePlayerColor(hex: String?) {
        val player = currentPlayer ?: return
        viewModelScope.launch {
            try {
                firestoreRepository.updatePlayerColor(gameId, player.id, hex)
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

    fun updateCommanderName(name: String?) {
        val player = currentPlayer ?: return
        viewModelScope.launch {
            try {
                firestoreRepository.updateCommanderName(gameId, player.id, name)
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

    fun updateGameSettings(maxPlayers: Int? = null, startingLife: Int? = null, gameMode: String? = null, maxTraitorRarity: String? = null) {
        if (!isHost) return
        viewModelScope.launch {
            try {
                cloudFunctionsRepository.updateGameSettings(gameId, maxPlayers, startingLife, gameMode, maxTraitorRarity)
                // The game doc flow echoes the other settings back, but rarity
                // is tracked locally (see _maxTraitorRarity above).
                maxTraitorRarity?.let { value ->
                    Rarity.fromValue(value)?.let { _maxTraitorRarity.value = it }
                }
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

}
